### PR TITLE
Feature/stage out

### PR DESCRIPTION
**Pull request by andre-merzky**
This adds output staging to EnTK's `AppManager` (similar to input staging via the `shared_data` attribute).  It's a  feature required for ReplicaExchange.

Sorry for not opening this PR earlier, this is ready for review/merge since quite while...

### CODE REVIEW

**Comment by andre-merzky** [comment]
@lee: can you please take care of this before the release (which should go out actually)?  @mturilli : any other suggestion for a reviewer?

File: src/radical/entk/execman/base/resource_manager.py [hunk 1]
```diff
@@ -139,6 +139,20 @@ def shared_data(self, data_list):
         self._shared_data = data_list
```
Comment (mturilli) [inline src/radical/entk/execman/base/resource_manager.py]: Do we have a `setter` for `shared_data`? If so, should we update the docstring of `shared_data` adding a setter?

File: src/radical/entk/execman/base/resource_manager.py [hunk 2]
```diff
@@ -139,6 +139,20 @@ def shared_data(self, data_list):
...
 
 
+    @property
+    def outputs(self):
+        """
+        :getter:    list of files to be staged from remote after execution
+        """
```
Comment (mturilli) [inline src/radical/entk/execman/base/resource_manager.py]: Do we have a setter for outputs too?

File: src/radical/entk/execman/base/resource_manager.py [hunk 3]
```diff
@@ -139,6 +139,20 @@ def shared_data(self, data_list):
...
+        return self._outputs
+
+
+    @shared_data.setter
```
Comment (mturilli) [inline src/radical/entk/execman/base/resource_manager.py]: For readibility and consistency, should we separate getters and setters as done in `appmanager`? I would try not to make the code grow 'organically'.

File: src/radical/entk/execman/rp/resource_manager.py [hunk 4]
```diff
@@ -159,6 +159,8 @@ def _pilot_state_cb(pilot, state):
                        'access_schema' : self._access_schema,
                        'queue'         : self._queue,
                        'cleanup'       : cleanup,
+                       'input_staging' : self._shared_data,
```
Comment (mturilli) [inline src/radical/entk/execman/rp/resource_manager.py]: This surprised me. Why `input_staging` is always also `shared_data`?

File: src/radical/entk/execman/rp/resource_manager.py [hunk 5]
```diff
@@ -171,24 +173,14 @@ def _pilot_state_cb(pilot, state):
 
             self._prof.prof('rreq submitted', uid=self._uid)
 
-            shared_staging_directives = list()
```
Comment (mturilli) [inline src/radical/entk/execman/rp/resource_manager.py]: Why this and the following is now not required?

File: src/radical/entk/execman/base/resource_manager.py (hunk 3 above)
Comment (lee212) [inline src/radical/entk/execman/base/resource_manager.py]: `@outputs.setter` it is ?

File: src/radical/entk/execman/rp/resource_manager.py (hunk 5 above)
Comment (lee212) [reply src/radical/entk/execman/rp/resource_manager.py]: ` 'input_staging' : self._shared_data` will replace the `stage_in` step.

File: src/radical/entk/appman/appmanager.py [hunk 6]
```diff
@@ -405,13 +438,15 @@ def run(self):
...
 
-        except Exception as ex:
-
-            self._logger.exception('Error in AppManager: %s' % ex)
+            self._logger.exception('Error in AppManager')
             self.terminate()
             raise
 
+        # return list of fetched output data, or None.
+        return ret
```
[the remark this answers is not in the document]
Comment (lee212) [reply src/radical/entk/appman/appmanager.py]: added `outputs()` to ret at 45cbee9

**Pull request merged**